File: src/lib/gamification.ts
import { datesLoggedInMonth, loggingStreak, spentInMonth } from './analytics.ts'
import { daysInMonth, monthKey } from './dates.ts'
import { formatMoney, niceAmount } from './money.ts'
import { collectAlerts } from './thresholds.ts'
import type {
  Achievement,
  AchievementId,
  AppState,
  LevelProgress,
  MonthPoint,
  Quest,
  StretchGoal,
} from '../types.ts'

export function stretchGoalForMonth(history: MonthPoint[], month: string): StretchGoal {
  const recent = history.filter((point) => point.month < month).slice(-3)

  if (recent.length === 0) {
    return {
      month,
      target: 200,
      baseline: 0,
      reason: 'First month in Grove — stash €200 and start the savings habit.',
    }
  }

  const last = recent[recent.length - 1]
  const average = recent.reduce((sum, point) => sum + point.net, 0) / recent.length
  const best = Math.max(...recent.map((point) => point.net))

  if (last.net <= 0 && average <= 0) {
    return {
      month,
      target: 80,
      baseline: last.net,
      reason: 'A clean reset: finish this month €80 in the black and rebuild momentum.',
    }
  }

  const reference = Math.max(last.net, average, 0)
  let target = reference * 1.1

  if (last.net > 0) {
    target = Math.max(target, last.net * 1.08)
    target = Math.min(target, last.net * 1.2)
  }

  if (last.net > 0 && best > average * 1.35 && last.net === best && average > 0) {
    target = last.net * 1.05
  }

  target = Math.max(50, niceAmount(target))

  const extra = Math.max(0, target - Math.max(last.net, 0))
  const perDay = extra / daysInMonth(month)
  const extraText =
    extra > 0
      ? ` That’s about ${formatMoney(perDay)} extra per day versus last month.`
      : ''

  return {
    month,
    target,
    baseline: niceAmount(reference),
    reason: `Recent savings sit around ${formatMoney(reference)}. Stretch to ${formatMoney(target)} this month.${extraText}`,
  }
}

export function monthQuests(input: {
  state: AppState
  month: string
  today: string
  stretch: StretchGoal
  monthPoint: MonthPoint
}): Quest[] {
  const { state, month, today, stretch, monthPoint } = input
  const through = monthKey(today) === month ? today : `${month}-31`
  const logged = datesLoggedInMonth(state.transactions, month, through)
  const logTarget = Math.min(10, Math.max(6, Math.ceil(daysInMonth(month) / 4)))
  const alerts = collectAlerts(state, month, today)
  const warningCount = alerts.filter((alert) => alert.level !== 'pace').length
  const overall = state.budgets.find((budget) => budget.categoryId === 'overall')
  const overallSpent = spentInMonth(state.transactions, month)
  const overallLimit = overall?.monthlyLimit ?? 0

  const stretchProgress = Math.max(0, monthPoint.net)

  return [
    {
      id: 'stretch',
      title: 'Stretch save',
      detail: `Put aside ${formatMoney(stretch.target)} after expenses.`,
      xp: 120,
      current: stretchProgress,
      target: stretch.target,
      unit: 'money',
      complete: monthPoint.net >= stretch.target,
    },
    {
      id: 'under-budget',
      title: 'Stay inside the fence',
      detail: overall
        ? `Keep overall spending at or under ${formatMoney(overall.monthlyLimit)}.`
        : 'Set an overall budget to activate this quest.',
      xp: 60,
      current: overallSpent,
      target: overallLimit || 1,
      unit: 'money',
      complete: Boolean(overall) && overallSpent <= overallLimit,
    },
    {
      id: 'calm-categories',
      title: 'Quiet categories',
      detail: 'Trigger no warning or breach on category budgets.',
      xp: 50,
      current: warningCount === 0 ? 1 : 0,
      target: 1,
      unit: 'count',
      complete: warningCount === 0,
    },
    {
      id: 'keep-logging',
      title: 'Keep the ledger warm',
      detail: `Log activity on ${logTarget} different days this month.`,
      xp: 40,
      current: logged,
      target: logTarget,
      unit: 'count',
      complete: logged >= logTarget,
    },
  ]
}

export function stretchHits(history: MonthPoint[]): { month: string; hit: boolean; goal: StretchGoal; net: number }[] {
  return history.map((point, index) => {
    const goal = stretchGoalForMonth(history.slice(0, index), point.month)
    return { month: point.month, hit: point.net >= goal.target, goal, net: point.net }
  })
}

export function collectAchievements(input: {
  state: AppState
  today: string
  history: MonthPoint[]
  hits: { month: string; hit: boolean }[]
}): Achievement[] {
  const { state, today, history, hits } = input
  const streak = loggingStreak(state.transactions, today)
  const completedMonths = history.filter((point) => point.month < monthKey(today))
  const overall = state.budgets.find((budget) => budget.categoryId === 'overall')
  const underBudgetMonths = completedMonths.filter((point) => {
    const spent = spentInMonth(state.transactions, point.month)
    return overall ? spent <= overall.monthlyLimit : point.net > 0
  })
  const quietMonths = completedMonths.filter((point) => {
    const monthEnd = `${point.month}-28`
    const alerts = collectAlerts(state, point.month, monthEnd)
    return alerts.filter((alert) => alert.level !== 'pace').length === 0
  })
  const balancedMonths = completedMonths.filter((point) => {
    const monthEnd = `${point.month}-28`
    const alerts = collectAlerts(state, point.month, monthEnd).filter(
      (alert) => alert.categoryId !== 'overall' && alert.level !== 'pace',
    )
    return alerts.length === 0
  })
  const hitMonths = hits.filter((hit) => hit.hit)
  let longestStretchRun = 0
  let run = 0
  for (const hit of hits) {
    run = hit.hit ? run + 1 : 0
    longestStretchRun = Math.max(longestStretchRun, run)
  }
  const cameBack = hits.some((hit, index) => index > 0 && hit.hit && !hits[index - 1].hit)
  const earlyRoot = state.transactions.some((transaction) => transaction.date.endsWith('-01'))
  const unlocked = new Set<AchievementId>()

  if (state.transactions.length > 0) unlocked.add('first-leaf')
  if (streak >= 7) unlocked.add('week-streak')
  if (underBudgetMonths.length > 0) unlocked.add('month-under')
  if (hitMonths.length > 0) unlocked.add('stretch-rookie')
  if (longestStretchRun >= 3) unlocked.add('stretch-streak')
  if (history.some((point) => point.net >= 500)) unlocked.add('big-save')
  if (quietMonths.length > 0) unlocked.add('quiet-month')
  if (balancedMonths.length > 0) unlocked.add('balanced-grove')
  if (cameBack) unlocked.add('comeback')
  if (earlyRoot) unlocked.add('early-root')

  return achievementCatalog().map((achievement) => ({
    ...achievement,
    unlocked: unlocked.has(achievement.id),
  }))
}

export function achievementCatalog(): Omit<Achievement, 'unlocked'>[] {
  return [
    { id: 'first-leaf', title: 'First leaf', detail: 'Log your first transaction.', xp: 15 },
    { id: 'early-root', title: 'Early root', detail: 'Log something on the first day of a month.', xp: 20 },
    { id: 'week-streak', title: 'Seven suns', detail: 'Keep a 7-day logging streak.', xp: 40 },
    { id: 'month-under', title: 'Inside the fence', detail: 'Finish a month under your overall budget.', xp: 50 },
    { id: 'quiet-month', title: 'Quiet canopy', detail: 'Finish a month with no warning thresholds tripped.', xp: 55 },
    { id: 'balanced-grove', title: 'Balanced grove', detail: 'Keep every category budget calm for a month.', xp: 60 },
    { id: 'stretch-rookie', title: 'Stretch sprout', detail: 'Hit a monthly stretch savings goal.', xp: 70 },
    { id: 'stretch-streak', title: 'Triple stretch', detail: 'Hit stretch goals three months in a row.', xp: 120 },
    { id: 'big-save', title: 'Deep roots', detail: 'Save at least €500 in a single month.', xp: 80 },
    { id: 'comeback', title: 'Second spring', detail: 'Hit a stretch goal after missing one.', xp: 45 },
  ]
}

export function computeXp(input: {
  transactionCount: number
  uniqueDays: number
  hits: { hit: boolean }[]
  underBudgetMonths: number
  questsComplete: number
  achievements: Achievement[]
}): number {
  const stretchXp = input.hits.filter((hit) => hit.hit).length * 120
  const achievementXp = input.achievements
    .filter((achievement) => achievement.unlocked)
    .reduce((sum, achievement) => sum + achievement.xp, 0)
  return (
    input.transactionCount * 6 +
    input.uniqueDays * 8 +
    stretchXp +
    input.underBudgetMonths * 40 +
    input.questsComplete * 20 +
    achievementXp
  )
}

export function levelFromXp(totalXp: number): LevelProgress {
  let remaining = Math.max(0, totalXp)
  let level = 1
  let need = 120
  while (remaining >= need) {
    remaining -= need
    level += 1
    need = Math.round(need * 1.18)
  }
  return {
    level,
    intoLevel: remaining,
    toNext: need,
    progress: need === 0 ? 1 : remaining / need,
    totalXp,
  }
}

export function pastUnderBudgetCount(state: AppState, history: MonthPoint[], today: string): number {
  const overall = state.budgets.find((budget) => budget.categoryId === 'overall')
  return history.filter((point) => {
    if (point.month >= monthKey(today)) return false
    const spent = spentInMonth(state.transactions, point.month)
    return overall ? spent <= overall.monthlyLimit : false
  }).length
}

File: lib/services/gamification.dart
import '../models/budget_limit.dart';
import '../models/gamification.dart';
import '../models/notice.dart';
import '../models/transaction.dart';
import '../ui/theme.dart';
import 'dates.dart';
import 'thresholds.dart';

class GameEngine {
  const GameEngine();

  StretchGoal stretchGoalForMonth(List<MonthPoint> history, String month) {
    final older =
        history.where((point) => point.month.compareTo(month) < 0).toList();
    final recent = older.length <= 3 ? older : older.sublist(older.length - 3);

    if (recent.isEmpty) {
      return StretchGoal(
        month: month,
        target: 200,
        baseline: 0,
        reason:
            'Pirmas mėnuo programėlėje — atidėkite 200 € ir pradėkite taupymo įprotį.',
      );
    }

    final last = recent.last;
    final average =
        recent.fold<double>(0, (sum, point) => sum + point.net) / recent.length;
    final best = recent.map((point) => point.net).reduce((a, b) => a > b ? a : b);

    if (last.net <= 0 && average <= 0) {
      return StretchGoal(
        month: month,
        target: 80,
        baseline: last.net,
        reason:
            'Švarus startas: užbaikite mėnesį 80 € pliusu ir atgaukite pagreitį.',
      );
    }

    final reference = [last.net, average, 0.0].reduce((a, b) => a > b ? a : b);
    var target = reference * 1.1;

    if (last.net > 0) {
      final floor = last.net * 1.08;
      final cap = last.net * 1.2;
      if (target < floor) target = floor;
      if (target > cap) target = cap;
    }

    if (last.net > 0 &&
        best > average * 1.35 &&
        last.net == best &&
        average > 0) {
      target = last.net * 1.05;
    }

    target = target < 50 ? 50 : niceAmount(target);

    final extra = target - (last.net > 0 ? last.net : 0);
    final extraClamped = extra < 0 ? 0.0 : extra;
    final perDay = extraClamped / daysInMonth(month);
    final extraText = extraClamped > 0
        ? ' Tai apie ${formatEur(perDay)} papildomai per dieną palyginti su praėjusiu mėnesiu.'
        : '';

    return StretchGoal(
      month: month,
      target: target,
      baseline: niceAmount(reference),
      reason:
          'Pastarųjų mėnesių taupymas apie ${formatEur(reference)}. Iššūkis — ${formatEur(target)} šį mėnesį.$extraText',
    );
  }

  List<Quest> monthQuests({
    required List<MoneyTx> txs,
    required List<BudgetLimit> budgets,
    required String month,
    required DateTime today,
    required StretchGoal stretch,
    required MonthPoint monthPoint,
    required List<ThresholdAlert> alerts,
  }) {
    final through =
        monthKey(today) == month ? dateKey(today) : '$month-31';
    final logged = datesLoggedInMonth(txs, month, through);
    final logTarget = (daysInMonth(month) / 4).ceil().clamp(6, 10);
    final warningCount =
        alerts.where((alert) => alert.level != AlertLevel.pace).length;
    BudgetLimit? overall;
    for (final budget in budgets) {
      if (budget.categoryId == 'overall') {
        overall = budget;
        break;
      }
    }
    final overallSpent = const ThresholdEngine().spentInMonth(txs, month);
    final overallLimit = overall?.monthlyLimit ?? 0;
    final stretchProgress = monthPoint.net < 0 ? 0.0 : monthPoint.net;

    return [
      Quest(
        id: 'stretch',
        title: 'Sutaupyk daugiau',
        detail: 'Po išlaidų atidėkite ${formatEur(stretch.target)}.',
        xp: 120,
        current: stretchProgress,
        target: stretch.target,
        unit: 'money',
        complete: monthPoint.net >= stretch.target,
      ),
      Quest(
        id: 'under-budget',
        title: 'Tilpk į biudžetą',
        detail: overall == null
            ? 'Nustatykite bendrą biudžetą, kad įjungtumėte šį iššūkį.'
            : 'Laikykite visas išlaidas iki ${formatEur(overall.monthlyLimit)}.',
        xp: 60,
        current: overallSpent,
        target: overallLimit == 0 ? 1 : overallLimit,
        unit: 'money',
        complete: overall != null && overallSpent <= overallLimit,
      ),
      Quest(
        id: 'calm-categories',
        title: 'Ramios kategorijos',
        detail: 'Nesužadinkite įspėjimo ar viršijimo nė vienoje kategorijoje.',
        xp: 50,
        current: warningCount == 0 ? 1 : 0,
        target: 1,
        unit: 'count',
        complete: warningCount == 0,
      ),
      Quest(
        id: 'keep-logging',
        title: 'Fiksuokite operacijas',
        detail: 'Turėkite aktyvumą $logTarget skirtingomis šio mėnesio dienomis.',
        xp: 40,
        current: logged.toDouble(),
        target: logTarget.toDouble(),
        unit: 'count',
        complete: logged >= logTarget,
      ),
    ];
  }

  List<StretchHit> stretchHits(List<MonthPoint> history) {
    return [
      for (var i = 0; i < history.length; i++)
        StretchHit(
          month: history[i].month,
          hit: history[i].net >=
              stretchGoalForMonth(history.sublist(0, i), history[i].month)
                  .target,
          goal: stretchGoalForMonth(history.sublist(0, i), history[i].month),
          net: history[i].net,
        ),
    ];
  }

  List<Achievement> collectAchievements({
    required List<MoneyTx> txs,
    required List<BudgetLimit> budgets,
    required DateTime today,
    required List<MonthPoint> history,
    required List<StretchHit> hits,
  }) {
    final streak = loggingStreak(txs, today);
    final completedMonths = history
        .where((point) => point.month.compareTo(monthKey(today)) < 0)
        .toList();
    BudgetLimit? overall;
    for (final budget in budgets) {
      if (budget.categoryId == 'overall') {
        overall = budget;
        break;
      }
    }
    final engine = const ThresholdEngine();
    final underBudgetMonths = completedMonths.where((point) {
      final spent = engine.spentInMonth(txs, point.month);
      return overall != null ? spent <= overall.monthlyLimit : point.net > 0;
    });
    final quietMonths = completedMonths.where((point) {
      final monthEnd = DateTime(
        int.parse(point.month.split('-')[0]),
        int.parse(point.month.split('-')[1]),
        28,
      );
      final alerts = engine.collectAlerts(
        txs: txs,
        budgets: budgets,
        month: point.month,
        today: monthEnd,
      );
      return alerts.where((alert) => alert.level != AlertLevel.pace).isEmpty;
    });
    final balancedMonths = completedMonths.where((point) {
      final monthEnd = DateTime(
        int.parse(point.month.split('-')[0]),
        int.parse(point.month.split('-')[1]),
        28,
      );
      final alerts = engine
          .collectAlerts(
            txs: txs,
            budgets: budgets,
            month: point.month,
            today: monthEnd,
          )
          .where(
            (alert) =>
                alert.categoryId != 'overall' && alert.level != AlertLevel.pace,
          );
      return alerts.isEmpty;
    });
    final hitMonths = hits.where((hit) => hit.hit);
    var longestStretchRun = 0;
    var run = 0;
    for (final hit in hits) {
      run = hit.hit ? run + 1 : 0;
      if (run > longestStretchRun) longestStretchRun = run;
    }
    var cameBack = false;
    for (var i = 1; i < hits.length; i++) {
      if (hits[i].hit && !hits[i - 1].hit) cameBack = true;
    }
    final earlyRoot = txs.any((tx) => dateKey(tx.bookedAt).endsWith('-01'));
    final unlocked = <String>{};
    if (txs.isNotEmpty) unlocked.add('first-leaf');
    if (streak >= 7) unlocked.add('week-streak');
    if (underBudgetMonths.isNotEmpty) unlocked.add('month-under');
    if (hitMonths.isNotEmpty) unlocked.add('stretch-rookie');
    if (longestStretchRun >= 3) unlocked.add('stretch-streak');
    if (history.any((point) => point.net >= 500)) unlocked.add('big-save');
    if (quietMonths.isNotEmpty) unlocked.add('quiet-month');
    if (balancedMonths.isNotEmpty) unlocked.add('balanced-grove');
    if (cameBack) unlocked.add('comeback');
    if (earlyRoot) unlocked.add('early-root');

    return [
      for (final item in achievementCatalog())
        item.copyWith(unlocked: unlocked.contains(item.id)),
    ];
  }

  List<Achievement> achievementCatalog() {
    return const [
      Achievement(
        id: 'first-leaf',
        title: 'Pirmas lapas',
        detail: 'Įrašykite pirmą operaciją.',
        xp: 15,
        unlocked: false,
      ),
      Achievement(
        id: 'early-root',
        title: 'Ankstyva šaknis',
        detail: 'Turėkite operaciją pirmą mėnesio dieną.',
        xp: 20,
        unlocked: false,
      ),
      Achievement(
        id: 'week-streak',
        title: 'Septynios saulės',
        detail: '7 dienų operacijų serija.',
        xp: 40,
        unlocked: false,
      ),
      Achievement(
        id: 'month-under',
        title: 'Viduje tvoros',
        detail: 'Užbaikite mėnesį neviršydami bendro biudžeto.',
        xp: 50,
        unlocked: false,
      ),
      Achievement(
        id: 'quiet-month',
        title: 'Ramus lajas',
        detail: 'Užbaikite mėnesį be įspėjimo ribų.',
        xp: 55,
        unlocked: false,
      ),
      Achievement(
        id: 'balanced-grove',
        title: 'Subalansuotas miškas',
        detail: 'Visą mėnesį ramios visų kategorijų ribos.',
        xp: 60,
        unlocked: false,
      ),
      Achievement(
        id: 'stretch-rookie',
        title: 'Taupymo daigas',
        detail: 'Įvykdykite mėnesio taupymo iššūkį.',
        xp: 70,
        unlocked: false,
      ),
      Achievement(
        id: 'stretch-streak',
        title: 'Trigubas iššūkis',
        detail: 'Tris mėnesius iš eilės pasiekite taupymo tikslą.',
        xp: 120,
        unlocked: false,
      ),
      Achievement(
        id: 'big-save',
        title: 'Gilios šaknys',
        detail: 'Sutaupykite bent 500 € per vieną mėnesį.',
        xp: 80,
        unlocked: false,
      ),
      Achievement(
        id: 'comeback',
        title: 'Antras pavasaris',
        detail: 'Pasiekite tikslą po praleisto mėnesio.',
        xp: 45,
        unlocked: false,
      ),
    ];
  }

  int computeXp({
    required int transactionCount,
    required int uniqueDays,
    required List<StretchHit> hits,
    required int underBudgetMonths,
    required int questsComplete,
    required List<Achievement> achievements,
  }) {
    final stretchXp = hits.where((hit) => hit.hit).length * 120;
    final achievementXp = achievements
        .where((item) => item.unlocked)
        .fold<int>(0, (sum, item) => sum + item.xp);
    return transactionCount * 6 +
        uniqueDays * 8 +
        stretchXp +
        underBudgetMonths * 40 +
        questsComplete * 20 +
        achievementXp;
  }

  LevelProgress levelFromXp(int totalXp) {
    var remaining = totalXp < 0 ? 0 : totalXp;
    var level = 1;
    var need = 120;
    while (remaining >= need) {
      remaining -= need;
      level += 1;
      need = (need * 1.18).round();
    }
    return LevelProgress(
      level: level,
      intoLevel: remaining,
      toNext: need,
      progress: need == 0 ? 1 : remaining / need,
      totalXp: totalXp,
    );
  }

  int pastUnderBudgetCount({
    required List<MoneyTx> txs,
    required List<BudgetLimit> budgets,
    required List<MonthPoint> history,
    required DateTime today,
  }) {
    BudgetLimit? overall;
    for (final budget in budgets) {
      if (budget.categoryId == 'overall') {
        overall = budget;
        break;
      }
    }
    if (overall == null) return 0;
    final engine = const ThresholdEngine();
    return history.where((point) {
      if (point.month.compareTo(monthKey(today)) >= 0) return false;
      final spent = engine.spentInMonth(txs, point.month);
      return spent <= overall!.monthlyLimit;
    }).length;
  }

  int datesLoggedInMonth(List<MoneyTx> txs, String month, String throughDate) {
    return txs
        .where(
          (tx) =>
              monthKey(tx.bookedAt) == month &&
              dateKey(tx.bookedAt).compareTo(throughDate) <= 0,
        )
        .map((tx) => dateKey(tx.bookedAt))
        .toSet()
        .length;
  }

  int loggingStreak(List<MoneyTx> txs, DateTime today) {
    final dates = txs.map((tx) => dateKey(tx.bookedAt)).toSet();
    var streak = 0;
    var cursor = DateTime(today.year, today.month, today.day);
    if (!dates.contains(dateKey(cursor))) {
      cursor = cursor.subtract(const Duration(days: 1));
    }
    while (dates.contains(dateKey(cursor))) {
      streak += 1;
      cursor = cursor.subtract(const Duration(days: 1));
    }
    return streak;
  }
}
